enhancements/ocp-coreos-layering/ocp-coreos-layering.md

```diff
@@ -0,0 +1,730 @@
+  (In contrast today, the existing `machine-os-content` has an ostree repository inside a UBI image, which is hard to inspect and cannot be used for derived builds).  For more information, see [ostree-rs-ext](https://github.com/ostreedev/ostree-rs-ext/) and [CoreOS layering](https://github.com/coreos/enhancements/pull/7).
+- This new format container will *also* be shipped alongside the toplevel release image with a matching version scheme. For example, there will now be e.g `quay.io/openshift-release-dev/rhel-coreos-8:4.12.0-rc.5-x86_64`.
+- The MCO will be modified to allow a cluster admin to override the `osImageURL` field on MachineConfigs to supply a new base image that overrides the default (RHEL) CoreOS image. The administrator can use the above image tags in a `Dockerfile` or other container build system.
+- Everything else largely stays the same; the MCD continues to perform config changes per node. (See: [Buildless layering support](#buildless-layering-support)). 
+
+#### Phase 1
+- Initial [preflight checks](#preflight-checks) are implemented. However, they will only be a warning and will not block image rollout since we will not (yet) have a way to bypass them.
+- Iterate on the mechanism established in Phase 0 by enabling rendered MachineConfigs to be layered into a final OS image before they are applied to a nodes’ disk by using the BuildController. (see [Create Final Pool Images](#create-final-pool-images)).
+- The BuildController will be reliant upon the OpenShift Image Builder. Alternatively, we could create our own custom solution (see: [How The Final Image Will Be Built](#how-the-final-image-will-be-built) for caveats).)
+
+#### Phase 2
+- CRD implementation is finalized. This enables the following:
+  - A cluster admin can build and stage a final image prior to rolling it out.
+  - Preflight checks can block an image rollout (assuming the documented overrides are not used).
+  - Allows us to remove our dependency on the in-cluster registry and/or ImageStreams since it gives the cluster admin a place to configure where the final image is pushed. This allows them to optionally use an in-cluster registry or ImageStream assuming they’ve enabled those features on their cluster.
+- Modify the BuildController to use customized build pods instead of OpenShift Image Builder to create the finalized OS image (assuming we did the initial implementation with OpenShift Image Builder).
+- Add support for heterogeneous clusters using BuildController.
+
+#### Phase 3
+- oc command is augmented with sub-commands to provide syntactic sugar and a more polished synchronous user experience for cluster admins around staging a new OS image, watching preflight checks pass, and rolling it out. The web UI gains an equivalent UX flow.
+
+**Note:** Rest of the enhancement proposal will focus on. For other phases, details can be found in [Future Phases](#future-phases) section.
+
+### Workflow Description
```

Comment:
Can this be moved up under Phase 0?

Reply:
As per enhancement template https://github.com/openshift/enhancements/blob/master/guidelines/enhancement_template.md#workflow-description  is a separate section inside proposal.